static/app/utils/useAsync.tsx

```diff
@@ -0,0 +1,382 @@
+  onRequestSuccess = _data => {},
+  onRequestError = (_error, _args) => {},
+}: Options): Result {
+  const api = useApi();
+  const location = useLocation();
+  const routes = useRoutes();
+  const params = useParams();
+
+  const [measurement, setMeasurement] = useState<MetricsState>({
+    hasMeasured: false,
+    finished: false,
+    error: false,
+  });
+
+  const initialState = {
+    data: {},
+    isLoading: false,
+    hasError: false,
+    isReloading: false,
+    errors: {},
+    remainingRequests: endpoints.length,
+  };
+  const [state, setState] = useState<State>({...initialState});
+
+  useEffect(() => {
+    const mount = async () => {
+      try {
+        await fetchData();
+      } catch (error) {
+        setState(prevState => ({...prevState, hasError: true}));
+        throw error;
+      }
+    };
+    if (routes && routes.length) {
+      metric.mark({name: `async-component-${getRouteStringFromRoutes(routes)}`});
+    }
+
+    if (reloadOnVisible) {
+      document.addEventListener('visibilitychange', visibilityReloader);
+    }
+
+    mount();
+
+    return () => {
+      // Anything in here is fired on component unmount.
+      api.clear();
+      document.removeEventListener('visibilitychange', visibilityReloader);
+    };
+  }, []);
+
+  useEffect(() => {
+    // Take a measurement from when this component is initially created until it finishes it's first
+    // set of API requests
+    if (!measurement.hasMeasured && measurement.finished && routes) {
+      const routeString = getRouteStringFromRoutes(routes);
+      metric.measure({
+        name: 'app.component.async-component',
+        start: `async-component-${routeString}`,
+        data: {
+          route: routeString,
+          error: measurement.error,
+        },
+      });
+      setMeasurement({...measurement, hasMeasured: true});
+    }
+  }, [measurement]);
+
+  useEffect(() => {
+    remountComponent();
+  }, [location.search, location.state, params]);
+
+  useEffect(() => {
+    if (endpoints.length && state.remainingRequests === 0 && !state.hasError) {
+      onLoadAllEndpointsSuccess();
+    }
+  }, [state.remainingRequests, state.hasError]);
+
+  // Check if we should measure render time for this component
+  function markShouldMeasure({
+    remainingRequests,
+    error,
+  }: {error?: boolean; remainingRequests?: number} = {}) {
+    if (measurement.hasMeasured) {
+      setMeasurement({
+        ...measurement,
+        finished: remainingRequests === 0,
+        error: error || measurement.error,
+      });
+    }
+  }
+
+  function remountComponent() {
+    if (shouldReload) {
+      return async () => {
+        await reloadData();
+      };
+    }
+    setState({...initialState});
+    return async () => {
+      await fetchData();
+    };
+  }
+
+  function visibilityReloader() {
+    return shouldReloadOnVisible && !state.isLoading && !document.hidden && reloadData();
+  }
+
+  async function reloadData() {
```

Comment:
It's actually not async
```suggestion
 function reloadData() {
```